### PR TITLE
[Metricbeat] Add metricbeat module for App Search

### PR DESCRIPTION
**Pull request by ioanatia**
App Search has a health API endpoint that can be used starting from 7.3.2.

It provides various stats which can be used to monitor the health of the instance.
I've picked up a few which should be more relevant:

- `jvm` memory usage
- worker `queues` counts
- response times for both dashboard and API requests

### CODE REVIEW

**Comment by fearful-symmetry** [comment]
@jsoriano Would this be a good candidate for light modules? It seems like a straightforward HTTP fetch.

**Comment by jsoriano** [comment]
> @jsoriano Would this be a good candidate for light modules? It seems like a straightforward HTTP fetch.

Not sure, things like `queues["failed.count"] = len(failed)` cannot be done with light modules.

**Comment by cachedout** [comment]
@ioanatia Nice to see this! I wrote up a brief proposal to track the addition of these metrics in Stack Monitoring: https://github.com/elastic/kibana/issues/46218

File: metricbeat/docs/fields.asciidoc [hunk 1]
```diff
@@ -817,6 +818,278 @@ type: long
...
+
+
+type: long
+
+--
+
+*`appsearch.stats.jvm.memory_usage.heap_used.bytes`*::
++
+--
+Head used by the JVM in bytes.
```
Comment (cachedout) [inline metricbeat/docs/fields.asciidoc]: s/Head/Heap/

File: metricbeat/docs/fields.asciidoc [hunk 2]
```diff
@@ -817,6 +818,278 @@ type: long
...
+
+
+type: long
+
+--
+
+*`appsearch.stats.jvm.memory_usage.heap_committed.bytes`*::
++
+--
+Commited heap to the JVM in bytes.
```
Comment (cachedout) [inline metricbeat/docs/fields.asciidoc]: S/Commited/Committed/

File: metricbeat/module/appsearch/stats/data.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+)
+
+func eventMapping(input []byte) (common.MapStr, error) {
+	var data map[string]interface{}
+	err := json.Unmarshal(input, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	var queues = data["queues"].(map[string]interface{})
```
Comment (ycombinator) [inline metricbeat/module/appsearch/stats/data.go]: Could you add some error handling here and in the next statement as well? So something like this:

```
queues, ok := data["queues"].(map[string]interface{})
if !ok {
  return nil, errors.New("queues is not a map")
}
```

Thanks!

File: metricbeat/module/appsearch/stats/data.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+func eventMapping(input []byte) (common.MapStr, error) {
+	var data map[string]interface{}
+	err := json.Unmarshal(input, &data)
+	if err != nil {
+		return nil, err
+	}
+
+	var queues map[string]interface{}
+	var ok bool
+	queues, ok = data["queues"].(map[string]interface{})
```
Comment (ycombinator) [inline metricbeat/module/appsearch/stats/data.go]: You don't need to declare `queues` and `ok` explicitly. Instead you can just do this:

```go
queues, ok := data["queues"].(map[string]interface{})
```

Same with `failed` below.

**Comment by ycombinator** [comment]
@ioanatia I've been messing around with the timeouts and it doesn't seem to help much. It seems that, perhaps due to resource constraints, the App Search web server takes quite a long time to spin up and the docker compose up times out.

I can think of a couple of options to move forward on this PR:

1. Is there some way to start up only parts of App Search? For instance, I wonder if it would be possible not to start up the embedded Filebeat sub-process. Perhaps that might help reduce the startup time?

2. If (1) isn't possible, I'd be okay with removing the integration test from this PR for now and filing an issue to track adding it later. That would let us get this PR in while buying us some time to figure out the timeout problem.

3. I also wonder if @jsoriano might have some ideas here. I suspect he may have dealt with slow-starting services (in CI environments) before when writing modules for them.

**Comment by ycombinator** [comment]
> Also, @ycombinator, should this module be added into the x-pack directory?

We have other Elastic stack products' Metricbeat modules outside the x-pack directory so I think, for consistency (for users and developers), it's okay for the appserarch module to be outside the x-pack directory as well.

**Comment by ycombinator** [comment]
> I think that everything is ok regarding timeouts and slow starting services. The problem can be in how the host address is configured for the test, I have added some related suggestions for that.

It would be great if changing the host address configuration fixes the issue. However, I had temporarily added some commits to this PR to show the Docker logs on failure. From what I saw the HTTP server on port 3002 that app search starts up was not getting started within 10 minutes. BTW, this seemed to happen only on CI; on my local laptop it started up within a couple of minutes tops.

File: x-pack/metricbeat/tests/system/test_appsearch.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import metricbeat
+import unittest
+
+
+class Test(XPackTest):
+    COMPOSE_SERVICES = ['appsearch']
+    COMPOSE_TIMEOUT = 600
+
+    @unittest.skipUnless(metricbeat.INTEGRATION_TESTS, "integration test")
+    def test_status(self):
```
Comment (ycombinator) [inline x-pack/metricbeat/tests/system/test_appsearch.py]: Probably should rename this to `test_stats` since that's the metricset being tested.

**Comment by jsoriano** [comment]
jenkins, test this again please

**Pull request merged**